core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -225,6 +232,91 @@ protected void perform(RelOptRuleCall call, @Nullable Filter filter,
     call.transformTo(relBuilder.build());
   }
 
+  private static void inferJoinEqualConditions(List<RexNode> rexNodes, Join join) {
```

Comment:
I'd suggest to add a small javadoc stating the purpose of this new method.
Also, like the already existing auxiliary method `validateJoinFilters`, perhaps it should be protected and non-static, just in case rule's sub-classes want to refine it.

Reply:
revised.